pkg/ccl/changefeedccl/scheduled_changefeed.go

```diff
@@ -0,0 +1,854 @@
+		// Sanity check: recurrence must be specified.
+		return nil, errors.New("RECURRING clause required")
+	}
+	{
+		rec, err := exprEval.String(ctx, schedule.Recurrence)
+		if err != nil {
+			return nil, err
+		}
+		spec.recurrence = &rec
+	}
+
+	enterpriseCheckErr := utilccl.CheckEnterpriseEnabled(
+		p.ExecCfg().Settings, p.ExecCfg().NodeInfo.LogicalClusterID(),
+		scheduleChangefeedOp)
+	spec.isEnterpriseUser = enterpriseCheckErr == nil
+
+	if !spec.isEnterpriseUser {
+		// Cannot use SCHEDULED CHANGEFEED w/out enterprise license.
+		// TODO(yevgeniy): Consider allowing cloudstorage sinks.
+		return nil, enterpriseCheckErr
+	}
+
+	spec.scheduleOpts, err = exprEval.KVOptions(
+		ctx, schedule.ScheduleOptions, scheduledChangefeedOptionExpectValues,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	spec.createChangefeedOptions, err = exprEval.KVOptions(
+		ctx, schedule.CreateChangefeed.Options, changefeedvalidators.CreateOptionValidations,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	evaluatedSinkURI, err := exprEval.String(ctx, schedule.SinkURI)
+	if err != nil {
+		return nil, err
+	}
+	spec.evaluatedSinkURI = &evaluatedSinkURI
+
+	return spec, nil
+}
+
+func makeChangefeedSchedule(
+	env scheduledjobs.JobSchedulerEnv,
+	owner username.SQLUsername,
+	label string,
+	recurrence *schedulebase.ScheduleRecurrence,
+	details jobspb.ScheduleDetails,
+	changefeedNode *tree.CreateChangefeed,
+) (*jobs.ScheduledJob, error) {
+	sj := jobs.NewScheduledJob(env)
+	sj.SetScheduleLabel(label)
+	sj.SetOwner(owner)
+
+	if err := sj.SetSchedule(recurrence.Cron); err != nil {
+		return nil, err
+	}
+
+	sj.SetScheduleDetails(details)
+
+	var args changefeedpb.ScheduledChangefeedExecutionArgs
+	args.ScheduledChangefeedStatement = tree.AsStringWithFlags(changefeedNode, tree.FmtParsable|tree.FmtShowPasswords)
+	any, err := pbtypes.MarshalAny(&args)
+	if err != nil {
+		return nil, err
+	}
+	sj.SetExecutionDetails(
+		tree.ScheduledChangefeedExecutor.InternalName(), jobspb.ExecutionArguments{Args: any},
+	)
+
+	return sj, nil
+}
+
+// dryRunChangefeed executes Changefeed in dry-run mode: we simply execute Changefeed
+// under transaction savepoint, and then rollback to that save point.
+func dryRunCreateChangefeed(
+	ctx context.Context,
+	p sql.PlanHookState,
+	scheduleID int64,
+	createChangefeedNode *tree.CreateChangefeed,
+) error {
+	sp, err := p.ExtendedEvalContext().Txn.CreateSavepoint(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = func() error {
+		annotated := &annotatedChangefeedStatement{
+			CreateChangefeed: createChangefeedNode,
+			CreatedByInfo: &jobs.CreatedByInfo{
+				Name: jobs.CreatedByScheduledJobs,
+				ID:   scheduleID,
+			},
+		}
+		changefeedFn, err := planCreateChangefeed(ctx, p, annotated)
+		if err != nil {
+			return err
+		}
+		return invokeCreateChangefeed(ctx, changefeedFn)
+	}()
+
+	if rollbackErr := p.ExtendedEvalContext().Txn.RollbackToSavepoint(ctx, sp); rollbackErr != nil {
+		return rollbackErr
```

Comment:
We probably should return errors.CombineErrors(rollbackErr, err) to make sure we preserve original error in case it's also non-nil